spec/index.md

```diff
@@ -1424,6 +1432,12 @@ The canister can access an argument. For `canister_init`, `canister_post_upgrade
 
     The reject message. Traps if there is no reject message (i.e. if `reject_code` is `0`).
 
+-   `ic0.msg_deadline : () -> i64`
+
+    The deadline, in nanoseconds since 1970-01-01, after which the caller will stop waiting for a response.
+
+    For calls with best-effort responses, the deadline is computed based on the time the call was made, and the `timeout_seconds` parameter provided by the caller. For other calls, a deadline of 0 will be returned.
```

Comment:
Did we consider the effects of unsynchronized clocks between subnets?

Reply:
As discussed in the meeting, monotonicity of time would be orthogonal to this PR (canisters can in theory observe time going back even without timeouts and best-effort responses).

Also as discussed, what would be relevant for timeouts would be large clock skews between subnets, because these would cause spurious request rejection (if the receiver subnet was >5 minutes ahead). Such rejection is technically still compliant with the spec, but canister communication would be effectively halted until the clocks are in reasonable sync again. We had an internal discussion on this and the conclusion was that we aren't worried about such a scenario. First, the canisters need to handle this situation anyways for other reasons (e.g., one of the subnets being under high load, or stalled), so such a situation shouldn't be a correctness problem, only an availability one. Second, given the existing time-related checks in our consensus mechanism, data about average quartz clock drifts, and the fact that we use NTP, we couldn't come up with a realistic scenario where we get large clock drift between subnets.

Reply:
> canisters can in theory observe time going back even without timeouts and best-effort responses

how?

Reply:
The question was about unsynchronized subnet clocks and time along the call graph; there, we don't guarantee time monotonicity across canisters on the same call graph.